### ISSUE TITLE
Design Suggestion: Reconsider eml.ml approach

### ISSUE THREAD

**Issue by sidkshatriya** [aantron/dream#156]
(Interestingly on the issue with ocaml-lsp not working well in Dream, someone else has already filed a ticket https://github.com/aantron/dream/issues/155 )

Thank you for this important and substantial open source contribution to the OCaml ecosystem! Thanks especially for the great documentation and many many examples too!

I had some comments on the `*.eml.ml` in the Dream framework:
- I personally place a lot of emphasis on the LSP server “understanding” everything in the ml file. I will gladly pay a small price in using a templating language that places nice with the OCaml LSP server. My guess is that a lot of beginners would prefer to have a LSP server that does not stop working in certain sections of their code – that is likely to be confusing and a net negative for them. Yes, using something like tyxml is a few extra steps but to me is a still worth the effort. The average OCaml beginner tends to be reasonably sophisticated and should be OK with tyxml. As you already would know – tyxml [%html] section allows variable interpolation and ocaml lsp server continues to work properly. Even in tyml jsx, ocaml-lsp-server works.
- While you gain some simplicity in source code using Dream specific templating you need to have a special dune rule to convert the eml.ml → ml which affects some of the beginner friendliness that I guess you’re aiming for

In brief, while I see your point in using the eml.ml approach I think the gains are mixed at best. I also notice that you have painstakingly pointed out that the eml.ml approach is purely optional and never compulsory anywhere. But the point is that this pattern is everywhere on the Dream website so it seems like it is the recommended approach.

My personal suggestion would be to put the tyxml approach front and center in the Dream framework examples. It plays nice with the ecosystem and does not add a “quirk” or unnecessary customization. People value cohesiveness with the broader OCaml ecosystem. Alternatively, dream framework can add its own tyxml like ppx.

What is your opinion about this design suggestion? Would be awesome to understand your opinion on this issue.

_BTW Its cool if you feel that the current approach makes sense. In that case please feel free to close this ticket. I just wanted to spark a discussion as I feel that it might be too late to make such a significant change in the future._

P.S. I also added this point on https://discuss.ocaml.org/t/dream-a-simple-yet-feature-complete-web-framework/7909/8?u=sid but that was an old-ish thread so its possible you haven't seen it.

**Comment by aantron** [comment]
> I personally place a lot of emphasis on the LSP server “understanding” everything in the ml file.

The main suggestion here in the short term is to use `.eml.html` files, as I also just suggested in the related issue https://github.com/aantron/dream/issues/155. I don't know yet how to make this more prominent in the examples, as `.eml.ml` files are just too convenient and concise not to use them in most examples :)

> The average OCaml beginner tends to be reasonably sophisticated and should be OK with tyxml.

I've had feedback both in agreement and not with this statement, including some from beginners that would have been intimidated if examples used mainly TyXML. The examples are also not aimed at an average OCaml beginner as at least I imagine them to be today. The examples are partially aimed at beginners that OCaml might have, if OCaml libraries were easier to use than they tend to be, and required less sophistication to get started with (I don't fully agree with the choice of "sophistication" for the meaning of my sentence here, it's a bit of a different meaning from that in the sentence I am quoting).

In addition, the examples are aimed at users like me (not beginners). I prefer to use "literal" templates to TyXML. Note that I am not in any way "against" TyXML, and have been a user of it and contributor to it as well.

> While you gain some simplicity in source code using Dream specific templating you need to have a special dune rule to convert the eml.ml → ml which affects some of the beginner friendliness that I guess you’re aiming for

This is true. In comparison with using TyXML, however, TyXML still requires a `(preprocess ...)` stanza. It's not as ugly as the eml rule, but it partially offsets the annoyance of editing `dune` files for eml.

**Comment by sidkshatriya** [comment]
Thanks for documenting your experience! 

I assume you're referring to https://ocsigen.org/tyxml/latest/manual/jsx right? Just so I understood you correctly, one would need to write your template code into a `.re` file, correct? Or are you envisaging another way of using the jsx ppx?

(And yes, the jsx syntax is very nice and intuitive. However I've only seen it be used in reason files).

**Comment by sidkshatriya** [comment]
@tmattio Thanks for detailed note.

> Unfortunately, it won't be as exhaustive as Merlin: only the first error of the file will be reported, as Dune (and the compiler) are only able to give the first error.

Can you give an example of this? Lets say I have a snippet of eml in my OCaml file. Right now what merlin does is that it gives me a lot of errors where the eml is but then "recovers" and does an acceptable job for the code following the eml snippet. This makes emls painful but not entirely a showstopper.

Once dune 3.0 comes out will this behavior regress? Will I continue getting helpful pointers from the the lsp later in my file? In other words, the new behavior of dune 3.0 + merlin is not clear to me in your note. Also isn't dune already integrated with merlin via the dot merlin reader?

**Comment by tmattio** [comment]
> Once dune 3.0 comes out will this behavior regress?

No, the integration with Dune 3.0 will only provide additional diagnostics (coming from the compiler), but will not interfere with the current Merlin diagnostics.

I'll let LSP maintainers communicate on this when they are ready, I just wanted to give some ideas of what the solutions could be to support `.eml` files in the editors going forward, and a heads up that the syntax will probably come in VSCode in the coming months 🙂

**Comment by faldor20** [comment]
Would it perhaps be possible to split the file before serving to the LSP/merlin? I believe this is how Razor pages in c# work as described here https://devblogs.microsoft.com/aspnet/new-experimental-razor-editor-for-visual-studio/

**Comment by faldor20** [comment]
I'm working on a solution to this, as far as language server support. Should have a prototype available soon.

**Comment by aantron** [comment]
There is now another templating alternative, [mlx](https://github.com/aantron/dream/tree/master/example/w-mlx#w-mlx), which should have LSP support.
